### PR TITLE
build: replace cross_build job with parallelized build matrix

### LINKED ISSUE #20638: Nightly builds report a confusing version
The metadata packaged with our nightly builds say it's version `2.0.0~rc.4`. This causes confusion when installing a nightly over an "official" release, because tools like `dpkg` report it as a downgrade.

Ideally, our nightly builds could report a realistic version (i.e. `2.0~nightly-${date}`) without requiring significant grunt work on our part to keep the goreleaser configs up-to-date.

### LINKED ISSUE #20855: Refactor CI pipeline to parallelize build/test for all target platforms instead of cross-compiling sequentially from linux/amd64
__Proposal:__
Rewrite our CI pipeline to split the process of building release binaries across multiple hosts, one per target platform. The control flow would be:
1. Top-level workflow is triggered when a tag is pushed. It runs on a cheap/generic linux executor.
2. Top-level workflow triggers following sub-workflows, then polls them to completion:
    * linux/amd64 build
    * linux/arm64 build
    * darwin/amd64 build
    * (Eventually) windows/amd64 build
    * (Eventually) darwin/arm64 build
3. After sub-workflows complete, top-level workflow collects output artifacts from each
4. Top-level workflow uploads collected artifacts to S3, builds & pushes Docker images(? unless we want this to run in the sub-workflows too)

__Current behavior:__
We currently use `goreleaser` to cross-compile for all our target platforms from a linux/amd64 host. Getting that setup to produce static binaries for our existing 3 targets (linux/amd64, linux/arm64, darwin/amd64) required us to:
* Build cross-compilers/libc for each target from source, and maintain the copy in our S3 bucket
* Shim layers of wrapper scripts into the CGo and Rust toolchains, to support swapping out the compiler/libc/build-tags used per-target

There are some major downsides to this setup:
* Release builds can't be parallelized, so our CI time grows with each target platform we add
* We only run unit tests against linux/amd64, so bugs could be lurking on other platforms without our knowledge
* Reproducing our release-build environment locally is very complex and resource-intensive

__Desired behavior:__
Building binaries for one of our target platforms should be as easy as:
1. Installing the basic tooling needed to run our `Makefile`
2. Installing a single Rust target (ideally the default for the host)
3. Running a single `go build` or `make` command

Builds should be able to use "native" libraries and tooling for each target, so we don't have to maintain copies build from source.

Builds should be parallelizable, so we don't have to worry about CI times exploding as we add support for more platforms.

It should be possible for us to run unit tests on all of our "Tier 1" platforms (list to be defined) before we release any bits.

### PR DESCRIPTION
**Pull request by danxmoran**
Closes #20855
Closes #20638

Instead of having separate build and cross_build jobs, we have a single build job that we run in a matrix across target OS and arch pairs. Jobs that only need the linux-amd64 build can start as soon as that build finishes.

We use goreleaser pro to handle the final package & upload steps of the pipeline. The pro license enables us to:
1. Import prebuilt binaries in our config
2. Split our config into multiple files, selectively importing/including the pieces we need for each build type
3. Define custom variables as a slightly more clear substitute for env vars

With these changes, the e2e tests are now the longest-running job in the pipeline. @wbaker85 suggested that porting the OSS e2e suite to use remocal might improve the situation there.

### CODE REVIEW

File: .circleci/config.yml
Comment (danxmoran) [inline .circleci/config.yml:150]: TODO before merge: un-comment this

File: .circleci/config.yml
Comment (lesam) [inline .circleci/config.yml:9]: Some version of this comment might still be useful

File: .circleci/config.yml
Comment (lesam) [inline .circleci/config.yml:39]: What's the status on that issue?

Comment (danxmoran) [reply .circleci/config.yml:39]: Still open, will need to be done before this can merge. I don't want to submit the request until this is approved since other merges will be blocked when the list of required jobs changes.

File: .circleci/config.yml
Comment (danxmoran) [reply .circleci/config.yml:9]: Restored the relevant bit

File: .circleci/config.yml
Comment (danxmoran) [inline .circleci/config.yml:99]: The mix of `hyphen-case` and `snake_case` conventions in the job names here has annoyed me for awhile - since we're going to request changing the list of required jobs anyways, I could also normalize the convention now. My preference would be to go with `hyphen-case` for everything

File: .circleci/config.yml
Comment (lesam) [reply .circleci/config.yml:39]: Oh I see - I was reading that other issue wrong, makes sense.

File: .circleci/config.yml
Comment (lesam) [reply .circleci/config.yml:99]: +100

Comment (danxmoran) [reply .circleci/config.yml]: Done

File: scripts/ci/build.sh
Comment (lesam) [inline scripts/ci/build.sh:52]: For CC - can we use `xcc.sh`?

If this is just to specify tags & ldflags, I'd prefer if we could just list the build commands for each arch instead of building them, it makes it easier to run the container and copy-paste a specific command to build (and iterate on changes etc)

File: scripts/ci/build-version.sh
Comment (lesam) [inline scripts/ci/build-version.sh:1]: Should this be a shared script in the build container?

File: .circleci/config.yml
Comment (lesam) [inline .circleci/config.yml:211]: I *love* the amount of deleted code here

File: scripts/ci/build.sh
Comment (danxmoran) [reply scripts/ci/build.sh:52]: Didn't realize `xcc.sh` was in the image, will update to use it & make the build commands more copy-paste friendly

**Review by lesam — COMMENTED** [review]
I love this, see a few comments / clarifications requested above.

What more work needs to be done to allow us to push a release tag to master? Looks like we're most of the way there?

**Comment by danxmoran** [comment]
> What more work needs to be done to allow us to push a release tag to master? Looks like we're most of the way there?

We'd need to:
1. Add the top-level workflow that triggers on the tag
2. Make sure we pass `build-type: release` to the build & package steps in the workflow
3. Add a `.goreleaser/release.yml` and any associated helper files (at least `.goreleaser/_uploads-release.yml`)

File: scripts/ci/build-version.sh
Comment (danxmoran) [reply scripts/ci/build-version.sh:1]: Sure, seems generic enough to be useful across projects

**Comment by lesam** [comment]
@danxmoran is there an issue already for release from master / if not can you create one with the above info?

**Comment by danxmoran** [comment]
Added a comment to #21551

File: .circleci/config.yml
Comment (wbaker85) [inline .circleci/config.yml:367]: I worry a little bit about running gorelease without the `--skip-validation` option since it ends up being kind of touchy about what things make the git checks fail. I do think its the right thing to do though, and hopefully as we port this to release branches (or adapt master to be able to do releases) we can keep the CI structured so that we are confident that if the nightly release passes the git checks, so too will a tagged release, so that there are no surprises when we try to do a tagged release.

File: .circleci/config.yml
Comment (lesam) [inline .circleci/config.yml:146]: LGTM once this comment is addressed

File: .circleci/config.yml
Comment (lesam) [inline .circleci/config.yml:632]: `VERSION="v$BRANCH"` makes the changelog `vmaster`, which is not great but is existing.

Comment (lesam) [reply .circleci/config.yml:632]: @wbaker85 fyi

File: .circleci/config.yml
Comment (lesam) [inline .circleci/config.yml:821]: Can we kill this script?

File: .circleci/config.yml
Comment (lesam) [reply .circleci/config.yml:146]: Though would still be nice to address the comments on the new scripts (copy-paste friendliness of build commands especially)

File: .circleci/config.yml
Comment (danxmoran) [reply .circleci/config.yml:821]: Yes

**Review by lesam — APPROVED** [review]

**Pull request merged**
